resources/schema.py

```diff
@@ -454,26 +458,68 @@ def mutate_and_get_payload(cls, root, info, **input):
         input["pier"] = get_node_from_global_id(
             info, input.pop("pier_id"), only_type=PierNode, nullable=False,
         )
-        input["berth_type"] = get_node_from_global_id(
-            info, input.pop("berth_type_id"), only_type=BerthTypeNode, nullable=False,
-        )
+        berth_type_id = input.pop("berth_type_id", None)
+        width = input.pop("width", None)
+        length = input.pop("length", None)
+        mooring_type = input.pop("mooring_type", None)
+        berth_type_params = [width, length, mooring_type]
+
+        # Cannot have both fields at the same time
+        if berth_type_id and any(berth_type_params):
+            raise VenepaikkaGraphQLError(_("Cannot receive BerthType and dimensions"))
+
+        if berth_type_id:
+            berth_type = get_node_from_global_id(
+                info, berth_type_id, only_type=BerthTypeNode, nullable=False,
+            )
+        elif all(berth_type_params):
+            berth_type, created = BerthType.objects.get_or_create(
+                width=width, length=length, mooring_type=mooring_type
+            )
+        else:
+            raise VenepaikkaGraphQLError(_("Missing fields to associate a BerthType"))
```

Comment:
seems like this logic could be in a separate function 🤔 `berth_type = get_berth_type_from_berth_mutation_input(input)` or something...